batch/utils/copy_files.py

```diff
@@ -0,0 +1,396 @@
+            paths = [(file, f'{dest}{os.path.basename(file)}', size)]
+        else:
+            paths = [(file, dest, size)]
+    elif src_paths:
+        if is_gcs_path(dest):
+            include_recurse_dir = dest.endswith('/') or next(await gcs_client.list_gs_files(dest, max_results=1), None) is not None
+        else:
+            include_recurse_dir = True
+        dest = dest.rstrip('/') + '/'
+        paths = [(src_path, dest + get_dest_path(src_path, src, include_recurse_dir), size) for src_path, size in src_paths]
+    else:
+        raise FileNotFoundError(src)
+
+    for src_path, dest_path, size in paths:
+        if is_gcs_path(src_path) and is_gcs_path(dest_path):
+            await file_pool.call(copy_file_within_gcs, copy_pool, src_path, dest_path, size)
+        elif is_gcs_path(src_path) and not is_gcs_path(dest_path):
+            await file_pool.call(read_file_from_gcs, copy_pool,
+                                 src_path, dest_path, size, max_partitions, min_partition_size)
+        elif not is_gcs_path(src_path) and is_gcs_path(dest_path):
+            await file_pool.call(write_file_to_gcs, copy_pool,
+                                 src_path, dest_path, size, max_partitions, min_partition_size)
+        else:
+            assert not is_gcs_path(src_path) and not is_gcs_path(dest_path)
+            await file_pool.call(copy_local_files, copy_pool, src_path, dest_path, size)
+
+
+async def main():
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument('--key-file', type=str, required=True)
+    parser.add_argument('--project', type=str, required=True)
+    parser.add_argument('--parallelism', type=int, default=10)
+    parser.add_argument('--concurrent-downloads', type=int, default=3)
+    parser.add_argument('--max-partitions', type=int, default=32)
+    parser.add_argument('--min-partition-size', type=str, default='256Mi')
+
+    args = parser.parse_args()
+
+    global thread_pool, gcs_client, file_lock_store
+
+    thread_pool = concurrent.futures.ThreadPoolExecutor()
+    credentials = google.oauth2.service_account.Credentials.from_service_account_file(args.key_file)
+    gcs_client = batch.google_storage.GCS(thread_pool, project=args.project, credentials=credentials)
+
+    file_worker_pool = AsyncWorkerPool(args.concurrent_downloads)
+    file_pool = WaitableSharedPool(file_worker_pool)
+
+    copy_worker_pool = AsyncWorkerPool(args.parallelism)
+    copy_pool = MultiWaitableSharedPool(copy_worker_pool)
+
+    file_lock_store = FileLockStore()
+
+    try:
+        coros = []
+        for line in sys.stdin:
+            src, dest = shlex.split(line.rstrip())
+            if '**' in src:
+                raise NotImplementedError(f'** not supported; got {src}')
+            coros.append(copies(file_pool, copy_pool, src, dest, args.max_partitions,
+                                parse_memory_in_bytes(args.min_partition_size)))
+        await asyncio.gather(*coros)
+        await file_pool.wait()
```

Comment:
why do we not have to wait on the copy_pool?

Reply:
Because it's implicitly waited on by coroutines in the file_pool.

Reply:
I added a wait for it now that I restructured the waitable pool